From eacd7d1b44b86ecb3b0e839315048a681d3e93d5 Mon Sep 17 00:00:00 2001
From: Alan Tull <atull@opensource.altera.com>
Date: Tue, 18 Oct 2016 09:18:18 -0500
Subject: [PATCH 017/151] FogBugz #411427: fpga: memory leaks in fpga manager
 debugfs

commit  63329981ae9a784a35a8483ca2a699c70c07d941 from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.64-lts

Fix a few places where buffers are allocated and not freed.

Signed-off-by: Alan Tull <atull@opensource.altera.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
Integrated-by: Jun Zhang <jun.zhang@windriver.com>
---
 drivers/fpga/fpga-mgr-debugfs.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/fpga/fpga-mgr-debugfs.c b/drivers/fpga/fpga-mgr-debugfs.c
index b25e337cb7ae..005360c3a509 100644
--- a/drivers/fpga/fpga-mgr-debugfs.c
+++ b/drivers/fpga/fpga-mgr-debugfs.c
@@ -41,13 +41,16 @@ static ssize_t fpga_mgr_firmware_write_file(struct file *file,
 	if (!buf)
 		return -ENOMEM;
 
-	if (copy_from_user(buf, user_buf, count))
+	if (copy_from_user(buf, user_buf, count)) {
+		kfree(buf);
 		return -EFAULT;
+	}
 
 	buf[count] = 0;
 	if (buf[count - 1] == '\n')
 		buf[count - 1] = 0;
 
+	/* Release previous firmware name (if any). Save current one. */
 	kfree(debugfs->firmware_name);
 	debugfs->firmware_name = buf;
 
@@ -109,8 +112,10 @@ static ssize_t fpga_mgr_image_write_file(struct file *file,
 	if (!buf)
 		return -ENOMEM;
 
-	if (copy_from_user(buf, user_buf, count))
+	if (copy_from_user(buf, user_buf, count)) {
+		kfree(buf);
 		return -EFAULT;
+	}
 
 	/* If firmware interface was previously used, forget it. */
 	kfree(debugfs->firmware_name);
@@ -121,6 +126,8 @@ static ssize_t fpga_mgr_image_write_file(struct file *file,
 		dev_err(&mgr->dev,
 		       "fpga_mgr_buf_load returned with value %d\n", ret);
 
+	kfree(buf);
+
 	return count;
 }
 
-- 
2.26.1

